File: src/intTest/java/io/javalin/ssl/plugin/TLSConfigTest.java
package io.javalin.ssl.plugin;

import io.javalin.Javalin;
import okhttp3.ConnectionSpec;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import javax.net.ssl.SSLHandshakeException;
import java.net.UnknownServiceException;
import java.util.Arrays;
import java.util.Collections;

import static org.junit.jupiter.api.Assertions.assertThrows;

@Tag("integration")
public class TLSConfigTest extends IntegrationTestClass {

    private static String[] substractArray(String[] array, String[] toRemove) {
        return Arrays.stream(array).filter(s -> !Arrays.asList(toRemove).contains(s)).toArray(String[]::new);
    }

    private static OkHttpClient clientWithTLSConfig(TLSConfig config) {

        ConnectionSpec spec = new ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
            .tlsVersions(config.getProtocols())
            .cipherSuites(config.getCipherSuites())
            .build();

        new TLSConfig(new String[]{""}, new String[]{""});
        return untrustedClientBuilder()
            .connectionSpecs(Collections.singletonList(spec))
            .build();
    }

    @Test
    void testModernConfigWithOldProtocols() {

        String[] protocols = substractArray(TLSConfig.OLD.getProtocols(), TLSConfig.MODERN.getProtocols()); // remove modern protocols from old protocols, so that ONLY unsupported protocols are left

        OkHttpClient client = clientWithTLSConfig(new TLSConfig(TLSConfig.MODERN.getCipherSuites(), protocols));

        int securePort = ports.getAndIncrement();
        String https = HTTPS_URL_WITH_PORT.apply(securePort);
        try (Javalin ignored = IntegrationTestClass.createTestApp(config -> {
            config.insecure = false;
            config.pemFromString(CERTIFICATE_AS_STRING, NON_ENCRYPTED_KEY_AS_STRING);
            config.securePort = securePort;
            config.tlsConfig = TLSConfig.MODERN;
        }).start()) {
            //Should fail with SSLHandshakeException because of the old protocols
            assertThrows(SSLHandshakeException.class, () -> client.newCall(new Request.Builder().url(https).build()).execute());
        }
    }

    @Test
    void testModernConfigWithOldCipherSuites() {

        String[] cipherSuites = substractArray(TLSConfig.OLD.getCipherSuites(), TLSConfig.MODERN.getCipherSuites()); // remove modern cipher suites from old cipher suites, so that we can test ONLY the old cipher suites

        OkHttpClient client = clientWithTLSConfig(new TLSConfig(cipherSuites, TLSConfig.MODERN.getProtocols()));

        int securePort = ports.getAndIncrement();
        String https = HTTPS_URL_WITH_PORT.apply(securePort);
        try (Javalin ignored = IntegrationTestClass.createTestApp(config -> {
            config.insecure = false;
            config.pemFromString(CERTIFICATE_AS_STRING, NON_ENCRYPTED_KEY_AS_STRING);
            config.securePort = securePort;
            config.tlsConfig = TLSConfig.MODERN;
        }).start()) {
            //Should fail with SSLHandshakeException because of the old cipher suites
            assertThrows(SSLHandshakeException.class, () -> client.newCall(new Request.Builder().url(https).build()).execute());
        }
    }

    @Test
    void testIntermediateConfigWithOldProtocols() {

        String[] protocols = substractArray(TLSConfig.OLD.getProtocols(), TLSConfig.INTERMEDIATE.getProtocols()); // remove intermediate protocols from old protocols, so that ONLY unsupported protocols are left

        OkHttpClient client = clientWithTLSConfig(new TLSConfig(TLSConfig.INTERMEDIATE.getCipherSuites(), protocols));

        int securePort = ports.getAndIncrement();
        String https = HTTPS_URL_WITH_PORT.apply(securePort);
        try (Javalin ignored = IntegrationTestClass.createTestApp(config -> {
            config.insecure = false;
            config.pemFromString(CERTIFICATE_AS_STRING, NON_ENCRYPTED_KEY_AS_STRING);
            config.securePort = securePort;
            config.tlsConfig = TLSConfig.INTERMEDIATE;
        }).start()) {
            //Should fail with SSLHandshakeException because of the old protocols
            assertThrows(UnknownServiceException.class, () -> client.newCall(new Request.Builder().url(https).build()).execute());
        }
    }


    @Test
    void testIntermediateConfigWithOldCipherSuites() {

        String[] cipherSuites = substractArray(TLSConfig.OLD.getCipherSuites(), TLSConfig.INTERMEDIATE.getCipherSuites()); // remove intermediate cipher suites from old cipher suites, so that we can test ONLY the old cipher suites

        OkHttpClient client = clientWithTLSConfig(new TLSConfig(cipherSuites, TLSConfig.INTERMEDIATE.getProtocols()));

        int securePort = ports.getAndIncrement();
        String https = HTTPS_URL_WITH_PORT.apply(securePort);
        try (Javalin ignored = IntegrationTestClass.createTestApp(config -> {
            config.insecure = false;
            config.pemFromString(CERTIFICATE_AS_STRING, NON_ENCRYPTED_KEY_AS_STRING);
            config.securePort = securePort;
            config.tlsConfig = TLSConfig.INTERMEDIATE;
        }).start()) {
            //Should fail with SSLHandshakeException because of the old cipher suites
            assertThrows(SSLHandshakeException.class, () -> client.newCall(new Request.Builder().url(https).build()).execute());
        }
    }

}

File: src/test/kotlin/io/javalin/community/ssl/TLSConfigTest.kt
package io.javalin.community.ssl

import io.javalin.community.ssl.certs.Server
import okhttp3.ConnectionSpec
import okhttp3.OkHttpClient
import okhttp3.Request
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import java.net.UnknownServiceException
import javax.net.ssl.SSLHandshakeException


@Tag("integration")
class TLSConfigTest : IntegrationTestClass() {
    @Test
    fun `test that a Modern TLS config does not allow old protocols`() {

        val protocols = TLSConfig.OLD.protocols.subtract(TLSConfig.MODERN.protocols.asIterable().toSet())

        // remove modern protocols from old protocols, so that ONLY unsupported protocols are left
        val client = clientWithTLSConfig(TLSConfig(TLSConfig.MODERN.cipherSuites, protocols.toTypedArray()))
        val securePort = ports.getAndIncrement()
        val https = HTTPS_URL_WITH_PORT.apply(securePort)
        createTestApp { config: SSLConfig ->
            config.insecure = false
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING)
            config.securePort = securePort
            config.tlsConfig = TLSConfig.MODERN
        }.start().let { _ ->
            //Should fail with SSLHandshakeException because of the old protocols
            Assertions.assertThrows(SSLHandshakeException::class.java) {
                client.newCall(
                    Request.Builder().url(
                        https
                    ).build()
                ).execute()
            }
        }
    }

    @Test
    fun `test that a Modern TLS config does not allow old cipher suites`() {
        val cipherSuites = TLSConfig.OLD.cipherSuites.subtract(TLSConfig.MODERN.cipherSuites.asIterable().toSet())
        // remove modern cipher suites from old cipher suites, so that we can test ONLY the old cipher suites
        val client = clientWithTLSConfig(TLSConfig(cipherSuites.toTypedArray(), TLSConfig.MODERN.protocols))
        val securePort = ports.getAndIncrement()
        val https = HTTPS_URL_WITH_PORT.apply(securePort)
        createTestApp { config: SSLConfig ->
            config.insecure = false
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING)
            config.securePort = securePort
            config.tlsConfig = TLSConfig.MODERN
        }.start().let { _ ->
            //Should fail with SSLHandshakeException because of the old cipher suites
            Assertions.assertThrows(SSLHandshakeException::class.java) {
                client.newCall(
                    Request.Builder().url(
                        https
                    ).build()
                ).execute()
            }
        }
    }

    @Test
    fun `test that an Intermediate TLS config does not allow old protocols`() {
        val protocols = TLSConfig.OLD.protocols.subtract(TLSConfig.INTERMEDIATE.protocols.asIterable().toSet())
        // remove intermediate protocols from old protocols, so that ONLY unsupported protocols are left
        val client = clientWithTLSConfig(TLSConfig(TLSConfig.INTERMEDIATE.cipherSuites, protocols.toTypedArray()))
        val securePort = ports.getAndIncrement()
        val https = HTTPS_URL_WITH_PORT.apply(securePort)
        createTestApp { config: SSLConfig ->
            config.insecure = false
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING)
            config.securePort = securePort
            config.tlsConfig = TLSConfig.INTERMEDIATE
        }.start().let { _ ->
            //Should fail with SSLHandshakeException because of the old protocols
            Assertions.assertThrows(UnknownServiceException::class.java) {
                client.newCall(
                    Request.Builder().url(https).build()
                ).execute()
            }
        }
    }

    @Test
    fun `test that an Intermediate TLS config does not allow old cipher suites`(){
        val cipherSuites = TLSConfig.OLD.cipherSuites.subtract(TLSConfig.INTERMEDIATE.cipherSuites.asIterable().toSet())
        // remove intermediate cipher suites from old cipher suites, so that we can test ONLY the old cipher suites
        val client = clientWithTLSConfig(TLSConfig(cipherSuites.toTypedArray(), TLSConfig.INTERMEDIATE.protocols))
        val securePort = ports.getAndIncrement()
        val https = HTTPS_URL_WITH_PORT.apply(securePort)
        createTestApp { config: SSLConfig ->
            config.insecure = false
            config.pemFromString(Server.CERTIFICATE_AS_STRING, Server.NON_ENCRYPTED_KEY_AS_STRING)
            config.securePort = securePort
            config.tlsConfig = TLSConfig.INTERMEDIATE
        }.start().let { _ ->
            //Should fail with SSLHandshakeException because of the old cipher suites
            Assertions.assertThrows(SSLHandshakeException::class.java) {
                client.newCall(
                    Request.Builder().url(
                        https
                    ).build()
                ).execute()
            }
        }
    }

    companion object {

        private fun clientWithTLSConfig(config: TLSConfig): OkHttpClient {


            val spec = ConnectionSpec.Builder(ConnectionSpec.MODERN_TLS)
                .cipherSuites(*config.cipherSuites)
                .tlsVersions(*config.protocols)
                .build()

            return listOf(spec).let {
                OkHttpClient.Builder()
                    .connectionSpecs(it)
                    .build()
            }
        }
    }
}
